# Preliminary Steps

In [52]:
import pandas as pd
import numpy as np
import re
import os

In [53]:
# Download and unzip shivamb/netflix-shows dataset
!kaggle datasets download shivamb/netflix-shows 

Dataset URL: https://www.kaggle.com/datasets/shivamb/netflix-shows
License(s): CC0-1.0
netflix-shows.zip: Skipping, found more recently modified local copy (use --force to force download)


In [54]:
# Only unzip if netflix-shows.zip hasn't been unzipped already
dir_files = os.listdir(os.getcwd())
unzipped_csvs = ['netflix-shows.csv']
if set(dir_files).isdisjoint(set(unzipped_csvs)) and "netflix-shows.zip" in dir_files:
    subprocess.run(["unzip", "-o", "netflix-shows.zip"], check=True)

Archive:  netflix-shows.zip
  inflating: netflix_titles.csv      


Try reading in the data on your own!

In [55]:
# TODO: Read your data as a pandas dataframe. Save the dataframe to a variable named "netflix"
DATA_PATH = os.path.join(os.getcwd(), "netflix_titles.csv")
netflix = pd.read_csv(DATA_PATH)
netflix

show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water              NaN   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans              NaN   
4         s5  TV Show           Kota Factory              NaN   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb              NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                   NaN            NaN   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                 ...            ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                                NaN            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

              date_added  release_year rating   duration  \
0     September 25, 2021          2020  PG-13     90 min   
1     September 24, 2021          2021  TV-MA  2 Seasons   
2     September 24, 2021          2021  TV-MA   1 Season   
3     September 24, 2021          2021  TV-MA   1 Season   
4     September 24, 2021          2021  TV-MA  2 Seasons   
...                  ...           ...    ...        ...   
8802   November 20, 2019          2007      R    158 min   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   
8804    November 1, 2019          2009      R     88 min   
8805    January 11, 2020          2006     PG     88 min   
8806       March 2, 2019          2015  TV-14    111 min   

                                              listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            description  
0     As her father nears the end of his life, filmm...  
1     After crossing paths at a party, a Cape Town t...  
2     To protect his family from a powerful drug lor...  
3     Feuds, flirtations and toilet talk go down amo...  
4     In a city of coaching centers known to train I...  
...                                                 ...  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...  

[8807 rows x 12 columns]

# EDA

We have a much larger dataset than last week. What are all the columns and what do they mean? What does each row mean? Take a some time to look through the data and understand what we're working with.

Some useful functions: describe(), unique(), sort_values(), dtypes(), shape, columns, info(), isnull(), value_counts(), and more if you can think of them. Get used to exploring data like this so you know what you're working with first!

In [56]:
# TODO: Do your EDA here!

netflix.describe()

release_year
count   8807.000000
mean    2014.180198
std        8.819312
min     1925.000000
25%     2013.000000
50%     2017.000000
75%     2019.000000
max     2021.000000

In [57]:
netflix.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


In [58]:
netflix.isna().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [59]:
netflix.dtypes

show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

In [60]:
netflix["country"].unique()

array(['United States', 'South Africa', nan, 'India',
       'United States, Ghana, Burkina Faso, United Kingdom, Germany, Ethiopia',
       'United Kingdom', 'Germany, Czech Republic', 'Mexico', 'Turkey',
       'Australia', 'United States, India, France', 'Finland',
       'China, Canada, United States',
       'South Africa, United States, Japan', 'Nigeria', 'Japan',
       'Spain, United States', 'France', 'Belgium',
       'United Kingdom, United States', 'United States, United Kingdom',
       'France, United States', 'South Korea', 'Spain',
       'United States, Singapore', 'United Kingdom, Australia, France',
       'United Kingdom, Australia, France, United States',
       'United States, Canada', 'Germany, United States',
       'South Africa, United States', 'United States, Mexico',
       'United States, Italy, France, Japan',
       'United States, Italy, Romania, United Kingdom',
       'Australia, United States', 'Argentina, Venezuela',
       'United States, United Kin

In [61]:
netflix["release_year"].value_counts()

release_year
2018    1147
2017    1032
2019    1030
2020     953
2016     902
        ... 
1959       1
1925       1
1961       1
1947       1
1966       1
Name: count, Length: 74, dtype: int64

# Granularity, Scope, and Temporality

Once you have completed your EDA, you should be able to answer each of these broad ideas about the data. If not, you can always do more EDA!

*Note, you can jot down some quick notes instead of a long answer to help remind yourself about certain characteristics of this dataframe

#### Granularity: What information does each row give us? How is each row unique (i.e., what identifies each row? This is called the **"primary key"**)

**Insert answer here**

Each row gives information about different movies, including their rating, genre, director, actors, year released, and more. Each row is uniquely identified by the show_id variable.

#### Scope: What are some interesting things we could learn from the data (we can find this out by exploring the columns we have)? Do we have to manipulate the data in some way to get what we're interested in?

**Insert answer here**

We can learn about the distribution of release dates and genres. Additionally, you can display how many movies/TV shows directors have made and/or actors have acted in.

#### Temporality: When was the data collected? How often is the data collected (if there is a pattern)? Do we need to adjust for consistency in the dates/times? Do we need to adjust the data type of our time variables?

**Insert answer here**

The date_added column should be a datetime object and not a string. Additionally, the release_date column should be an integer and not a string. It's hard to tell when the data was collected and how often since the selection of movies and TV shows range from the early 1900s to today.

We will be having people share their responses to each of these questions and discuss what everybody found in the data!

# Faithfulness

This is all about deciding whether you can trust the data in the form it came in, or whether you need to make adjustments to do so.

Did you find any strange or inconsistent values? Can you figure out how the data was collected? Are there any duplicate values (in this case, there shouldn't be because each movie is a separate one)?

**Insert answer here**

There are no duplicate values in the dataset. 

Possibly one of the most important questions is what to do with default values. These can be NaN, NA, 0, or something else entirely. Based on what we see from this data, what should we do with our missing values, and should we take different steps for different columns? You can check the slides for some examples of what people do with missing data.

In [62]:
netflix

show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water              NaN   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans              NaN   
4         s5  TV Show           Kota Factory              NaN   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb              NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                   NaN            NaN   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                 ...            ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                                NaN            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

              date_added  release_year rating   duration  \
0     September 25, 2021          2020  PG-13     90 min   
1     September 24, 2021          2021  TV-MA  2 Seasons   
2     September 24, 2021          2021  TV-MA   1 Season   
3     September 24, 2021          2021  TV-MA   1 Season   
4     September 24, 2021          2021  TV-MA  2 Seasons   
...                  ...           ...    ...        ...   
8802   November 20, 2019          2007      R    158 min   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   
8804    November 1, 2019          2009      R     88 min   
8805    January 11, 2020          2006     PG     88 min   
8806       March 2, 2019          2015  TV-14    111 min   

                                              listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            description  
0     As her father nears the end of his life, filmm...  
1     After crossing paths at a party, a Cape Town t...  
2     To protect his family from a powerful drug lor...  
3     Feuds, flirtations and toilet talk go down amo...  
4     In a city of coaching centers known to train I...  
...                                                 ...  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...  

[8807 rows x 12 columns]

In [63]:
netflix.isna().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

**Insert answer here**

The only columns that have NaN values are the following: director, cast, country, date_added, rating, and duration. For handling the NaN values, I think it would be optimal to drop the rows that have NaN values in just date_added, rating, and duration since they only make up a small fraction of the entire dataset. For the director, cast, and country NaN values, you could try filling in the NaN values using frequent values or random values in their respective column.

We'll discuss again afterwards!

# Variable Typing

Just for practice, it's important to be able to identify what type of variable something is just by looking at it. As a refresher, the 4 types are Quantitative Continous, Quantitative Discrete, Qualitative Ordinal, and Qualitative Nominal.

#### Question: What are the variables types of
a) Director

b) Country

c) Date Added

d) Release Year


**Insert Answer Here**

a) Director is Qualitative Nominal

b) Country is Qualitative Nominal

c) Date Added is Quantitative Discrete

d) Release Year is Quantitative Discrete

# Main Exercise

A common, but often misguided, way to handle datasets that have outliers/untrustworthy data is to drop all rows or columns that have missing values.


Find the shape of our dataset before we do any data manipulation and compare it with the shape after you use the dropna function. What do you notice and how many columns or rows were removed from our data?

Tip: You should make a deep copy of a dataframe by using {data name}.copy(deep=True). Give it an appropriate name to reflect the fact that this will be the dataframe that will have all null values dropped without any previous operations performed on it.

In [64]:
# TODO: Find the shape of our original data vs the shape of our data after we dropna
netflix_dropped = netflix.copy(deep=True)
netflix_dropped = netflix.dropna()

original_shape = netflix.shape
dropped_shape = netflix_dropped.shape

print(f"Original shape: {original_shape}")
print(f"Dropped shape: {dropped_shape}")

Original shape: (8807, 12)
Dropped shape: (5332, 12)


You should've found that some of our data was removed after using the drop_na function. Let's go back to our original dataset before we used the function.

Try creating a list of the columns that have null values. Feel free to search things up on StackOverflow, the Pandas documentations, etc.

In [65]:
# TODO: Create a list of columns that have null values

null_columns = netflix.columns[netflix.isna().any()].tolist()
null_columns

['director', 'cast', 'country', 'date_added', 'rating', 'duration']

Now that we have the columns where missing values are present, we have a better idea of what the data types of our missing values actually area.

Try filling in the missing director values with an empty string "". You can check if you did this operation correctly by checking the columns that have missing values and replacing the original dataframe name in the code with the name of the copy you should make for this task.

In [66]:
# TODO: Fill in missing director values with an emoty string

netflix_filled = netflix.copy(deep=True)

netflix_filled = netflix_filled.fillna({"director": ""})
netflix_filled.isna().sum()

show_id           0
type              0
title             0
director          0
cast            825
country         831
date_added       10
release_year      0
rating            4
duration          3
listed_in         0
description       0
dtype: int64

Let's check if filling the "director" column will help reduce the number of rows that will be removed when we use the dropna function.

Use the dropna function on the dataframe that is created after filling in the "director" column and find its shape. Compare the shape of this dataframe to the shape of the original dataframe and the dataframe after dropping all null values.

In [67]:
# TODO: Compare size of original data, data when dropping na after replacing missing directors with empty string, and data with just dropna

netflix_filled_dropped = netflix_filled.dropna()

print(f"Original shape: {original_shape}")
print(f"Dropped shape: {netflix_dropped.shape}")
print(f"Filled and dropped shape: {netflix_filled_dropped.shape}")

Original shape: (8807, 12)
Dropped shape: (5332, 12)
Filled and dropped shape: (7290, 12)


You should notice that the size of the dataframe that is created when we drop null values after filling in the "director" column will have less rows than the original dataframe, but more than the dataframe that is created when we just drop all values without performing any operations.

This should serve as a basic exercise on how you can approach data that is not 100% clean. Most data in consulting projects and in the real world will force you to find ways to balance the preservation of the original data and getting rid of the unecessary parts of the data.

### String Manipulation. We'll load in a new dataset for this part.

In [68]:
# TODO: load in the kazanova/sentiment140 data and unzip the file that is downloaded.

!kaggle datasets download kazanova/sentiment140

dir_files = os.listdir(os.getcwd())
unzipped_csvs = ['training.1600000.processed.noemoticon.csv']
if set(dir_files).isdisjoint(set(unzipped_csvs)) and "sentiment140.zip" in dir_files:
    subprocess.run(["unzip", "-o", "sentiment140.zip"], check=True)

Dataset URL: https://www.kaggle.com/datasets/kazanova/sentiment140
License(s): other
sentiment140.zip: Skipping, found more recently modified local copy (use --force to force download)


In [69]:
# TODO: Read the data from the csv file that is generated from unzipping as a dataframe.
# You should use the latin-1 encoding for this step. Name the dataframe tweets_data

DATA_PATH_TWEETS = os.path.join(os.getcwd(), "training.1600000.processed.noemoticon.csv")
tweets_data = pd.read_csv(DATA_PATH_TWEETS, encoding="latin-1")
tweets_data

0  1467810369  Mon Apr 06 22:19:45 PDT 2009  NO_QUERY  \
0        0  1467810672  Mon Apr 06 22:19:49 PDT 2009  NO_QUERY   
1        0  1467810917  Mon Apr 06 22:19:53 PDT 2009  NO_QUERY   
2        0  1467811184  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY   
3        0  1467811193  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY   
4        0  1467811372  Mon Apr 06 22:20:00 PDT 2009  NO_QUERY   
...     ..         ...                           ...       ...   
1599994  4  2193601966  Tue Jun 16 08:40:49 PDT 2009  NO_QUERY   
1599995  4  2193601969  Tue Jun 16 08:40:49 PDT 2009  NO_QUERY   
1599996  4  2193601991  Tue Jun 16 08:40:49 PDT 2009  NO_QUERY   
1599997  4  2193602064  Tue Jun 16 08:40:49 PDT 2009  NO_QUERY   
1599998  4  2193602129  Tue Jun 16 08:40:50 PDT 2009  NO_QUERY   

         _TheSpecialOne_  \
0          scotthamilton   
1               mattycus   
2                ElleCTF   
3                 Karoli   
4               joy_wolf   
...                  ...   
1599994  AmandaMarie1028   
1599995      TheWDBoards   
1599996           bpbabe   
1599997     tinydiamondz   
1599998   RyanTrevMorris   

        @switchfoot http://twitpic.com/2y1zl - Awww, that's a bummer.  You shoulda got David Carr of Third Day to do it. ;D  
0        is upset that he can't update his Facebook by ...                                                                   
1        @Kenichan I dived many times for the ball. Man...                                                                   
2          my whole body feels itchy and like its on fire                                                                    
3        @nationwideclass no, it's not behaving at all....                                                                   
4                            @Kwesidei not the whole crew                                                                    
...                                                    ...                                                                   
1599994  Just woke up. Having no school is the best fee...                                                                   
1599995  TheWDB.com - Very cool to hear old Walt interv...                                                                   
1599996  Are you ready for your MoJo Makeover? Ask me f...                                                                   
1599997  Happy 38th Birthday to my boo of alll time!!! ...                                                                   
1599998  happy #charitytuesday @theNSPCC @SparksCharity...                                                                   

[1599999 rows x 6 columns]

In [70]:
# Some cleanup of the data, this cell should run correctly if you named and read the dataframe properly
tweets = tweets_data.drop(columns = ["NO_QUERY"]).rename(columns={"0": "Polarity",
                                                                  "1467810369": "id",
                                                                  "Mon Apr 06 22:19:45 PDT 2009" : "Date",
                                                                  "_TheSpecialOne_" : "Username"})
final_tweets = tweets.set_axis([*tweets.columns[:-1], 'Tweet'], axis=1)
final_tweets.head(5)

Polarity          id                          Date       Username  \
0         0  1467810672  Mon Apr 06 22:19:49 PDT 2009  scotthamilton   
1         0  1467810917  Mon Apr 06 22:19:53 PDT 2009       mattycus   
2         0  1467811184  Mon Apr 06 22:19:57 PDT 2009        ElleCTF   
3         0  1467811193  Mon Apr 06 22:19:57 PDT 2009         Karoli   
4         0  1467811372  Mon Apr 06 22:20:00 PDT 2009       joy_wolf   

                                               Tweet  
0  is upset that he can't update his Facebook by ...  
1  @Kenichan I dived many times for the ball. Man...  
2    my whole body feels itchy and like its on fire   
3  @nationwideclass no, it's not behaving at all....  
4                      @Kwesidei not the whole crew

#### Now that we have our **final_tweets** dataset, we can start working with it. Let's practice string methods first.

In [71]:
# TODO: Make all of the strings in the "Username" columns lowercase

final_tweets["Username"] = final_tweets["Username"].str.lower()
final_tweets

Polarity          id                          Date         Username  \
0               0  1467810672  Mon Apr 06 22:19:49 PDT 2009    scotthamilton   
1               0  1467810917  Mon Apr 06 22:19:53 PDT 2009         mattycus   
2               0  1467811184  Mon Apr 06 22:19:57 PDT 2009          ellectf   
3               0  1467811193  Mon Apr 06 22:19:57 PDT 2009           karoli   
4               0  1467811372  Mon Apr 06 22:20:00 PDT 2009         joy_wolf   
...           ...         ...                           ...              ...   
1599994         4  2193601966  Tue Jun 16 08:40:49 PDT 2009  amandamarie1028   
1599995         4  2193601969  Tue Jun 16 08:40:49 PDT 2009      thewdboards   
1599996         4  2193601991  Tue Jun 16 08:40:49 PDT 2009           bpbabe   
1599997         4  2193602064  Tue Jun 16 08:40:49 PDT 2009     tinydiamondz   
1599998         4  2193602129  Tue Jun 16 08:40:50 PDT 2009   ryantrevmorris   

                                                     Tweet  
0        is upset that he can't update his Facebook by ...  
1        @Kenichan I dived many times for the ball. Man...  
2          my whole body feels itchy and like its on fire   
3        @nationwideclass no, it's not behaving at all....  
4                            @Kwesidei not the whole crew   
...                                                    ...  
1599994  Just woke up. Having no school is the best fee...  
1599995  TheWDB.com - Very cool to hear old Walt interv...  
1599996  Are you ready for your MoJo Makeover? Ask me f...  
1599997  Happy 38th Birthday to my boo of alll time!!! ...  
1599998  happy #charitytuesday @theNSPCC @SparksCharity...  

[1599999 rows x 5 columns]

#### Try making all the entries in the Tweet column uppercase and save this in a final_tweets_upper dataframe.

In [72]:
# TODO: make a copy of final_tweets and make the "tweets" column all uppercase

final_tweets_copy = final_tweets.copy(deep=True)
final_tweets_copy["Tweet"] = final_tweets_copy["Tweet"].str.upper()
final_tweets_copy

Polarity          id                          Date         Username  \
0               0  1467810672  Mon Apr 06 22:19:49 PDT 2009    scotthamilton   
1               0  1467810917  Mon Apr 06 22:19:53 PDT 2009         mattycus   
2               0  1467811184  Mon Apr 06 22:19:57 PDT 2009          ellectf   
3               0  1467811193  Mon Apr 06 22:19:57 PDT 2009           karoli   
4               0  1467811372  Mon Apr 06 22:20:00 PDT 2009         joy_wolf   
...           ...         ...                           ...              ...   
1599994         4  2193601966  Tue Jun 16 08:40:49 PDT 2009  amandamarie1028   
1599995         4  2193601969  Tue Jun 16 08:40:49 PDT 2009      thewdboards   
1599996         4  2193601991  Tue Jun 16 08:40:49 PDT 2009           bpbabe   
1599997         4  2193602064  Tue Jun 16 08:40:49 PDT 2009     tinydiamondz   
1599998         4  2193602129  Tue Jun 16 08:40:50 PDT 2009   ryantrevmorris   

                                                     Tweet  
0        IS UPSET THAT HE CAN'T UPDATE HIS FACEBOOK BY ...  
1        @KENICHAN I DIVED MANY TIMES FOR THE BALL. MAN...  
2          MY WHOLE BODY FEELS ITCHY AND LIKE ITS ON FIRE   
3        @NATIONWIDECLASS NO, IT'S NOT BEHAVING AT ALL....  
4                            @KWESIDEI NOT THE WHOLE CREW   
...                                                    ...  
1599994  JUST WOKE UP. HAVING NO SCHOOL IS THE BEST FEE...  
1599995  THEWDB.COM - VERY COOL TO HEAR OLD WALT INTERV...  
1599996  ARE YOU READY FOR YOUR MOJO MAKEOVER? ASK ME F...  
1599997  HAPPY 38TH BIRTHDAY TO MY BOO OF ALLL TIME!!! ...  
1599998  HAPPY #CHARITYTUESDAY @THENSPCC @SPARKSCHARITY...  

[1599999 rows x 5 columns]

#### Now let's try to replace something in a string!

In [73]:
# TODO: create a dataframe that filters out the final_tweets dataframe's tweets and only shows tweets with underscores

tweets_with_underscores = final_tweets_copy[final_tweets_copy["Tweet"].str.contains("_")]
tweets_with_underscores

Polarity          id                          Date         Username  \
7               0  1467811795  Mon Apr 06 22:20:05 PDT 2009  2hood4hollywood   
21              0  1467814119  Mon Apr 06 22:20:40 PDT 2009        cooliodoc   
68              0  1467825084  Mon Apr 06 22:23:30 PDT 2009    presidentsnow   
108             0  1467838188  Mon Apr 06 22:26:54 PDT 2009      jess_higley   
118             0  1467839586  Mon Apr 06 22:27:18 PDT 2009        sonyolmos   
...           ...         ...                           ...              ...   
1599986         4  2193579092  Tue Jun 16 08:38:58 PDT 2009        cathriiin   
1599987         4  2193579191  Tue Jun 16 08:38:59 PDT 2009          tellman   
1599988         4  2193579211  Tue Jun 16 08:38:59 PDT 2009        jazzstixx   
1599992         4  2193579477  Tue Jun 16 08:39:00 PDT 2009      chloeamisha   
1599993         4  2193579489  Tue Jun 16 08:39:00 PDT 2009        evolvetom   

                                                     Tweet  
7                     @TATIANA_K NOPE THEY DIDN'T HAVE IT   
21        @ANGRY_BARISTA I BAKED YOU A CAKE BUT I ATED IT   
68       @LT_ALGONQUIN AGREED, I SAW THE FAILWHALE ALLL...  
108      @MARYKATHERINE_Q I KNOW! I HEARD IT THIS AFTER...  
118      @ERRE_SC AAW I MISS YA ALL TOO.. IM LEAVING TO...  
...                                                    ...  
1599986  @LA_R_A NEVEER  I THINK THAT YOU BOTH WILL GET...  
1599987  @ROY_EVERITT HA- GOOD JOB. THAT'S RIGHT - WE G...  
1599988                 @MS_HIP_HOP IM GLAD UR DOING WELL   
1599992                                  @SCOOBY_GRITBOYS   
1599993  @CLIFF_FORSTER YEAH, THAT DOES WORK BETTER THA...  

[75329 rows x 5 columns]

Task 2: In the cell below, use .str.replace to replace all the underscores in Username with a space (" ") in the tweeters_with__ dataframe. Output the changed tweeters_with__ dataframe.

Doc: https://pandas.pydata.org/docs/reference/api/pandas.Series.str.replace.html

In [74]:
# TODO: replace all "_" with " " in the "Username" column

tweets_with_underscores["Username"] = tweets_with_underscores["Username"].str.replace("_", " ", regex=False)
tweets_with_underscores

/var/folders/79/yhqc0t4j20v0crtdf03zmpzm0000gn/T/ipykernel_17267/3059449471.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tweets_with_underscores["Username"] = tweets_with_underscores["Username"].str.replace("_", " ", regex=False)


Polarity          id                          Date         Username  \
7               0  1467811795  Mon Apr 06 22:20:05 PDT 2009  2hood4hollywood   
21              0  1467814119  Mon Apr 06 22:20:40 PDT 2009        cooliodoc   
68              0  1467825084  Mon Apr 06 22:23:30 PDT 2009    presidentsnow   
108             0  1467838188  Mon Apr 06 22:26:54 PDT 2009      jess higley   
118             0  1467839586  Mon Apr 06 22:27:18 PDT 2009        sonyolmos   
...           ...         ...                           ...              ...   
1599986         4  2193579092  Tue Jun 16 08:38:58 PDT 2009        cathriiin   
1599987         4  2193579191  Tue Jun 16 08:38:59 PDT 2009          tellman   
1599988         4  2193579211  Tue Jun 16 08:38:59 PDT 2009        jazzstixx   
1599992         4  2193579477  Tue Jun 16 08:39:00 PDT 2009      chloeamisha   
1599993         4  2193579489  Tue Jun 16 08:39:00 PDT 2009        evolvetom   

                                                     Tweet  
7                     @TATIANA_K NOPE THEY DIDN'T HAVE IT   
21        @ANGRY_BARISTA I BAKED YOU A CAKE BUT I ATED IT   
68       @LT_ALGONQUIN AGREED, I SAW THE FAILWHALE ALLL...  
108      @MARYKATHERINE_Q I KNOW! I HEARD IT THIS AFTER...  
118      @ERRE_SC AAW I MISS YA ALL TOO.. IM LEAVING TO...  
...                                                    ...  
1599986  @LA_R_A NEVEER  I THINK THAT YOU BOTH WILL GET...  
1599987  @ROY_EVERITT HA- GOOD JOB. THAT'S RIGHT - WE G...  
1599988                 @MS_HIP_HOP IM GLAD UR DOING WELL   
1599992                                  @SCOOBY_GRITBOYS   
1599993  @CLIFF_FORSTER YEAH, THAT DOES WORK BETTER THA...  

[75329 rows x 5 columns]

# Regex

---
### Part 1 — Does the title start with a number? 

Some Netflix titles start with a number, like `"13 Reasons Why"` or `"21 Jump Street"`.

**Task:** Use `str.contains()` with a regex pattern to find all titles that **start with a digit**. Store the result in a new DataFrame called `number_titles` and print how many there are.

**Hint:** Should be 130


In [75]:
# TODO: write a pattern that matches titles starting with a digit
pattern = r"^\d"
number_titles = netflix[netflix["title"].str.contains(pattern, na=False)]
print(f"Titles starting with a digit: {len(number_titles)}")
number_titles["title"].head(10)

Titles starting with a digit: 130


188                 2 Alone in Paris
323                          30 Rock
324                          44 Cats
404    9to5: The Story of a Movement
438                 2 Weeks in Lagos
558                        6 Bullets
774                         2 Hearts
850                         99 Songs
851                 99 Songs (Tamil)
852                99 Songs (Telugu)
Name: title, dtype: object

---
### Part 2 — Search descriptions for key words using `\w` and `[ ]`

**Task:** Find all titles whose description mentions the word `"love"` **or** `"romance"` (case-insensitive). Count them and print 5 examples.

Then, as a follow-up, find titles whose description contains **a number followed immediately by the word `"day"` or `"days"`** — e.g. `"30 days"` or `"1 day"`.

**Hints:**
- Pass `case=False` to `str.contains()` to ignore capitalisation.
- For the number + day pattern: `\d+` matches the number, `\s` matches the space, `days?` matches `"day"` or `"days"` (the `?` makes the `s` optional).
- Answer should be 826 and 9 respectively


In [76]:
# TODO: descriptions mentioning love or romance (case-insensitive)
pattern_love = r"love|romance"  # <-- your pattern here

love_titles = netflix[netflix["description"].str.contains(pattern_love, case=False, na=False)]
print(f"Titles mentioning love or romance: {len(love_titles)}")
love_titles["title"].head(5)

Titles mentioning love or romance: 826


24                   Jeans
25    Love on the Spectrum
26          Minsara Kanavu
27               Grown Ups
30         Ankahi Kahaniya
Name: title, dtype: object

In [77]:
# TODO: descriptions mentioning a number followed by "day" or "days"
pattern_days = r"\d+\sdays?" 

days_titles = netflix[netflix["description"].str.contains(pattern_days, case=False, na=False)]
print(f"Titles with 'N day(s)' in description: {len(days_titles)}")
days_titles[["title", "description"]].head(5)

Titles with 'N day(s)' in description: 9


title  \
1565                         Just The Way You Are   
2888                                Hi Bye, Mama!   
4829                             Sunday's Illness   
5031                                    Forgotten   
5893  Winter on Fire: Ukraine's Fight for Freedom   

                                            description  
1565  An overconfident teen bets he can make a homel...  
2888  When the ghost of a woman gains a second chanc...  
4829  Decades after being abandoned as a young child...  
5031  When his abducted brother returns seemingly a ...  
5893  Over 93 days in Ukraine, what started as peace...